File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def check(self,weights,mid):
        count=0
        sum1=0
        for i in range(len(weights)):
            sum1+=weights[i]
            if sum1>mid:
                count+=1
                sum1=weights[i]
        return count+1      
            
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low=max(weights)
        high=sum(weights)
        while low<high:
            mid=(low+high)//2
            #print(mid)
            #print(self.check(weights,mid))
            if self.check(weights,mid)<=days:
                high=mid
            else:
                low=mid+1
        return low      

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool solve(int mid,int days,vector<int> &weights){
        int count=0;
        int running_sum=0;
        for(int i=0;i<weights.size();i++){
            running_sum+=weights[i];
            if(running_sum>mid){
                count+=1;
                running_sum=weights[i];
            }
        }
        if (count+1<=days){
            return true;
        }
        else{
            return false;
        }
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int maxi=INT_MIN;
        int sum=0;
        int n=weights.size();
        for(int i=0;i<n;i++){
            maxi=max(maxi,weights[i]);
            sum+=weights[i];
        }
        int low=maxi;
        int high=sum;
        int ans=0;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(solve(mid,days,weights)){
                ans=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;
    }
};